### ISSUE TITLE
Creating Netflix Like Nested Collection View

### ISSUE THREAD

**Issue by Smiller193** [Instagram/IGListKit#1309]
So I have used IGListKit before to create a comment section in my app and though it took some time it ultimately worked very well.

I was contemplating utilizing the same collection view to create a Netflix like collectionView that  is vertically scrollable as well as horizontally scrollable across each row.

I saw some post about utilizing nested adapters. I noticed there seemed to some errors when it came to the number of items which left my sort of skeptical.

My app is working with events like workout classes and 5k runs that occur in the area. Each event has a model like so

    class Event:NSObject{
        var key: String?
        let currentEventName: String
        let currentEventImage: String
        let currentEventPromo: String?
        let currentEventDescription: String
        //nested properties
        let currentEventStreetAddress: String
        let currentEventCity: String
        let currentEventState: String
        let currentEventDate: String?
        let currentEventEndDate: String?
        let currentEventTime: String?
        let currentEventEndTime: String?
        let eventPrice: String
        let endTime: Date
        let endTimeInSeconds: TimeInterval?
        let startTime: Date
        let startTimeInSeconds: TimeInterval?
    
        let currentEventZip: Int
        var category: String
        //nested properties stop
        var currentAttendCount: Int
        var isAttending = false
        var eventTags: [String]?
        var userHost: [String]?
        var orgHost: [String]?
    
        
        
        
        var eventDictionary: [String: Any]{
            let dateDict = ["start:date":currentEventDate, "start:time": currentEventTime,"end:time":currentEventEndTime, "end:date": currentEventEndDate]
            let timeDict = ["end":endTime, "start": startTime]
            
            return ["event:name":currentEventName,"event:imageURL" : currentEventImage,
                    "event:description": currentEventDescription, "attend:count": currentAttendCount,
                    "event:street:address": currentEventStreetAddress,"event:zip": currentEventZip,"event:price":eventPrice,
                    "event:state": currentEventState, "event:city": currentEventCity, "event:promo": currentEventPromo ?? "", "event:date": dateDict, "event:category":category,"event:datetime": timeDict]
        }
        
        //init for event submission
        init?(eventKey:String, currentEventName:String, currentEventImage:String,currentEventPromo:String,currentEventDescription:String,currentEventStreetAddress:String,currentEventCity:String,currentEventState: String,currentEventDate: String,currentEventEndDate: String,currentEventTime: String, currentEventEndTime: String,eventPrice: String,endTimeInSeconds: TimeInterval,startTimeInSeconds: TimeInterval,currentEventZip: Int,category: String,currentAttendCount: Int,eventTags: [String]?,userHost: [String]?) {
            
            self.key = eventKey
            self.currentEventName = currentEventName
            self.currentEventImage = currentEventImage
            self.currentEventPromo = currentEventPromo
            self.currentEventDescription = currentEventDescription
            self.currentEventStreetAddress = currentEventStreetAddress
            self.currentEventCity = currentEventCity
            self.currentEventState = currentEventState
            self.currentEventZip = currentEventZip
            self.currentEventDate = currentEventDate
            self.currentEventEndDate = currentEventEndDate
            self.currentEventTime = currentEventTime
            self.currentEventEndTime = currentEventEndTime
            self.startTimeInSeconds = startTimeInSeconds
            self.endTimeInSeconds = endTimeInSeconds
            self.eventPrice = eventPrice
            self.category = category
            self.currentAttendCount = currentAttendCount
            self.eventTags = eventTags
            self.userHost = userHost
            self.endTime = Date()
            self.startTime = Date()
            
        }
    
        
    }
    
    extension Event {
        static public func ==(rhs: Event, lhs: Event) ->Bool{
            return rhs.key == lhs.key
        }
    }
    
    
    extension Event: ListDiffable{
        public func diffIdentifier() -> NSObjectProtocol {
            return key! as NSObjectProtocol
        }
        public func isEqual(toDiffableObject object: ListDiffable?) ->Bool{
            guard let object = object as? Event else {
                return false
            }
            return  self.key==object.key
        }
    }

However each event can either be featured or not featured. Regardless of whether it is featured or not it goes under a collection based off of its category (City, Nightlife, Sporting Events, Concert, etc). Regular events are further broken up into smaller subcategories for better organization (Concert -> hip hop, rock and roll, jazz, etc).  I use a dictionary of arrays to easily categorize them and pass them to the screen based first off of the main category (City, Nightlife, Sporting Events, Concert, etc) and then sub categories which may fall under those bigger categories (Concert -> hip hop, rock and roll, jazz, etc). Each of these events are displayed using a collectionView with cells that have collection views in them so that they are horizontally scrolling.

I am facing somewhat of an issue on proper implementation of the same UI approach using IGListKit. 
For example 

       // 1 objects(for:) returns an array of data objects that should show up in the collection view. 
        func objects(for listAdapter: ListAdapter) -> [ListDiffable] {
            <#code#>
        }

This code takes an array of objects that show up in the collection view which typically isn't difficult. The way I was passing the events to there respective nested collection Views before was based off of the section. The top section got the featured events and the bottom section received all the regular events.  I also used two arrays, What would be an easier way to place all the events in one single List Diffable array to be used here? Im thinking enums but im not really sure.

My second bout of confusion is here 

    func listAdapter(_ listAdapter: ListAdapter, sectionControllerFor object: Any) -> ListSectionController {
        return ListSectionController()
    }

Typically a section controller seems to be some sort of a wrapper for a CollectionViewCell that handles all the underlying methods (numberOfItems, didSelectItemAt, sizeForItemAt, etc). From what I can remember this creates a section controller for each item in the object list. Would I create a section controller for featuredEvents and a section controller for regular events. I want featured events to have one section because the featured events will be one row at the top that contains all the featured events for a specific category and the regular events may contain more than one section depending on the number of sub categories there are (Concert -> hip hop, rock and roll, etc)

@rnystrom @weyert

**Comment by Smiller193** [comment]
@weyert @bdotdub @weyert @zpao @levi

**Comment by DimaVartanian** [comment]
It's a little tough to unpack some of this because there's a lot here but I'll give you an idea of how we usually approach setups like this, with nested collections.

As you said, typically one section controller is typically meant to represent a single "object". Worth mentioning that a single object does not necessarily mean a single cell (A single section controller/section can have multiple cells, that might represent different pieces of the same object, to be displayed together).

I would design a system like this with nested UICollectionViews (with their own list adapters and IGListKit setups).

The top level collection view represents your top level vertically scrolling list. I would create a discrete object for each section which would have an array of the items for that section and then any additional metadata (such as a title, subtitle, whether it's "featured", etc).

The data source for that collection view would then create a section controller which would represent each horizontally scrolling list. One of the cells in those section controllers would contain another collection view, where each section controller would represent your lowest level item (the actual individual events).

Basically, it would help to create a clear abstraction at each level which would clearly separate responsibilities and let you customize everything as needed. Intermediate models to represent the top level sections will also make things easier.
